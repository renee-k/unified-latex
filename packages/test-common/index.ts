import { expect } from "vitest";
import { processLatexToAstViaUnified } from "@unified-latex/unified-latex";
import { parseMinimal } from "@unified-latex/unified-latex-util-parse";
import { VFile } from "unified-lint-rule/lib";
import * as Ast from "../unified-latex-types";
import { trimRenderInfo as _trimRenderInfo } from "../unified-latex-util-render-info";

declare global {
    namespace jest {
        interface Matchers<R> {
            toFormatAs(obj: any, formatter: any): R;
        }
    }
}

expect.extend({
    async toFormatAs(inStr, outStr, formatter) {
        if (typeof formatter !== "function") {
            throw new Error(
                "Must pass in a formatting function as the second argument when using `toFormatAs`"
            );
        }
        const formatted = await formatter(inStr);

        const pass = this.equals(formatted, outStr);

        return {
            pass,
            message: () =>
                `When formatting\n\n${this.utils.EXPECTED_COLOR(
                    inStr
                )}\n\nthe output did ${
                    pass ? "" : "not"
                } format correctly\n\n${this.utils.diff(outStr, formatted)}`,
        };
    },
});

/**
 * Parse a string directly into an `Ast.Node[]` array.
 */
export function strToNodes(str: string, skipTrimRenderInfo = false) {
    let value: string | undefined;
    let file: VFile | undefined;
    value = str;
    file = processLatexToAstViaUnified().processSync({ value });
    if (!skipTrimRenderInfo) {
        const root = _trimRenderInfo(file.result as any) as Ast.Root;
        return root.content;
    }
    return (file.result as Ast.Root).content;
}

/**
 * Parse a string directly into an `Ast.Node[]` array but don't trim
 * any whitespace.
 */
export function strToNodesRaw(str: string) {
    let file: VFile | undefined;
    file = processLatexToAstViaUnified().processSync({ value: `{${str}}` });
    const root = _trimRenderInfo(file.result as any) as Ast.Root;
    return (root.content[0] as Ast.Group).content;
}

/**
 * Parse a string directly into an `Ast.Node[]` array without any fancy reparsing/argument attaching, etc.
 */
export function strToNodesMinimal(str: string, skipTrimRenderInfo = false) {
    const parsed = parseMinimal(str);
    if (!skipTrimRenderInfo) {
        const root = _trimRenderInfo(parsed) as Ast.Root;
        return root.content;
    }
    return parsed.content;
}
