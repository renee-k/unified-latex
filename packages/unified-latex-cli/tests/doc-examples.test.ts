/**
 * Rus the files in `../../examples`. This test need not be in this directory, but it was a convenient place to put it (2024-02-19).
 */
import { describe, it, expect } from "vitest";
import util from "util";
import "../../test-common";
import { exec as _exec } from "node:child_process";
import * as path from "node:path";
import spawn from "cross-spawn";
import glob from "glob";

/* eslint-env jest */

// Make console.log pretty-print by default
const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

describe(
    "unified-latex-doc-examples",
    () => {
        const examplesPath = path.resolve(__dirname, "../../../examples");
        const exampleFiles = glob.sync(`${examplesPath}/*.ts`);
        for (const exampleFile of exampleFiles) {
            it(`example ${exampleFile.split("/").pop()}`, async () => {
                const stdout = await executeCommand(`npx`, [
                    "vite-node",
                    exampleFile,
                ]);
                expect(stdout).toMatchSnapshot();
            });
        }
    },
    {
        timeout: 60 * 1000,
    }
);

/**
 * Run commands with arguments using "cross-spawn", which correctly escapes arguments
 * so that end results are the same across different shells.
 */
function executeCommand(
    executablePath: string,
    args: string[]
): Promise<string> {
    return new Promise((resolve, reject) => {
        const childProcess = spawn(executablePath, args, { stdio: "pipe" });

        let stdoutData = "";

        childProcess.stdout!.on("data", (data) => {
            stdoutData += data.toString();
        });

        childProcess.on("error", (err) => {
            reject(err);
        });

        childProcess.on("close", (code) => {
            if (code === 0) {
                resolve(stdoutData);
            } else {
                reject(new Error(`Child process exited with code ${code}`));
            }
        });
    });
}
