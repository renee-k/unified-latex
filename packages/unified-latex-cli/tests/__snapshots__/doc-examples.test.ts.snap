// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unified-latex-doc-examples > example count-macros.ts 1`] = `
"Macro statistics:

All macros: {
  textbf: [ '\\\\\\\\textbf{an}', '\\\\\\\\textbf{amazing}' ],
  LaTeX: [ '\\\\\\\\LaTeX' ],
  textit: [ '\\\\\\\\textit{some}', '\\\\\\\\textit{\\\\\\\\textbf{amazing}}' ],
  '^': [ '^{x}', '^{\\\\\\\\infty}', '^{n}' ],
  sum: [ '\\\\\\\\sum' ],
  _: [ '_{n=0}' ],
  infty: [ '\\\\\\\\infty' ],
  frac: [ '\\\\\\\\frac{x^{n}}{n!}' ]
}
Math mode macros: [ '^', 'sum', '_', '^', 'infty', 'frac', '^' ]
"
`;

exports[`unified-latex-doc-examples > example custom-macros.ts 1`] = `
"\\\\textbf{custom} \\\\abc
\\\\textbf{custom} \\\\abc{macro}
\\\\textbf \\\\abc{macro}  
"
`;

exports[`unified-latex-doc-examples > example expanding-or-replacing-macros.ts 1`] = `
"\\\\newcommand{\\\\abc}[1]{ABC} My \\\\textbf{custom} \\\\abc{macro}.
 My \\\\textbf{custom} ABC.
\\\\abc{fun} $x=ABC$
 My \\\\textbf{custom} ABC.
 My \\\\textbf{custom} ABC.
"
`;

exports[`unified-latex-doc-examples > example ignore-defaults.ts 1`] = `
"\\\\textbf \\\\abc
^
\\\\color{blue} \\\\textbf {foo}
"
`;

exports[`unified-latex-doc-examples > example using-unified.ts 1`] = `
"{
  type: 'macro',
  content: 'textbf',
  position: {
    source: undefined,
    start: { offset: 9, line: 2, column: 9 },
    end: { offset: 16, line: 2, column: 16 }
  },
  _renderInfo: { inParMode: true },
  args: [
    {
      type: 'argument',
      content: [Array],
      openMark: '{',
      closeMark: '}'
    }
  ]
}
{
  type: 'macro',
  content: 'emph',
  position: {
    source: undefined,
    start: { offset: 9, line: 2, column: 9 },
    end: { offset: 16, line: 2, column: 16 }
  },
  _renderInfo: { inParMode: true },
  args: [
    {
      type: 'argument',
      content: [Array],
      openMark: '{',
      closeMark: '}'
    }
  ]
}
This is \\\\emph{an} example of a \\\\LaTeX{} document with \\\\textit{some} macros.
\\\\[
	e^{x} = \\\\sum_{n=0}^{\\\\infty}\\\\frac{x^{n}}{n!}.
\\\\]
What an \\\\textit{\\\\emph{amazing}} formula!
{
  type: 'macro',
  content: 'emph',
  position: {
    source: undefined,
    start: { offset: 9, line: 2, column: 9 },
    end: { offset: 16, line: 2, column: 16 }
  },
  _renderInfo: { inParMode: true },
  args: [
    {
      type: 'argument',
      content: [Array],
      openMark: '{',
      closeMark: '}'
    }
  ]
}
"
`;
